pandas/core/internals/blocks.py

```diff
@@ -739,7 +739,8 @@ def to_native_types(self, slicer=None, na_rep='nan', quoting=None,
         mask = isna(values)
 
         if not self.is_object and not quoting:
-            values = values.astype(str)
+            itemsize = writers.max_len_string(na_rep)
```

Comment:
maybe a better way is to do the ``values[mask] = na_rep`` before doing what we were doing originally which is the ``.astype(str)`` (which I think is pretty fast).

you prob need to least the ``values[mask] = na_rep`` in the else block though.

Reply:
Oh, I think it could work. 

>you prob need to least the `values[mask] = na_rep` in the else block though.

I moved the `values[mask] = na_rep` before the `if/else`, then left the `values.astype(str)` and `np.array(values, dtype='object')` to take care of the typing. It appears to have worked, at least from the couple of unit test classes I had open in the IDE for testing.

Running `test_fast.sh` now, and if that works, will revert the cython change, push to see if Travis is happy, and in parallel run the `asv` test again and report back its results.

Reply:
yes this is somewhat perf sensitive code, so IIRC that's why this has this kind of ordereing, but will see.

Reply:
@jreback 

The change I did was

```diff
diff --git a/pandas/core/internals/blocks.py b/pandas/core/internals/blocks.py
index 1bf122558..542925e88 100644
--- a/pandas/core/internals/blocks.py
+++ b/pandas/core/internals/blocks.py
@@ -738,13 +738,12 @@ class Block(PandasObject):
             values = values[:, slicer]
         mask = isna(values)
 
+        values[mask] = na_rep
         if not self.is_object and not quoting:
-            itemsize = writers.max_len_string(na_rep)
-            values = values.astype("<U{size}".format(size=itemsize))
+            values = values.astype(str)
         else:
             values = np.array(values, dtype='object')
 
-        values[mask] = na_rep
         return values
 
     # block actions ####
```

Resulted in

```
============================================= 95 failed, 45225 passed, 1185 skipped, 1000 xfailed, 22 xpassed, 121 warnings, 23 error in 440.68 seconds =============================================
```

Then tried this one

```diff
diff --git a/pandas/core/internals/blocks.py b/pandas/core/internals/blocks.py
index 1bf122558..0aa744071 100644
--- a/pandas/core/internals/blocks.py
+++ b/pandas/core/internals/blocks.py
@@ -739,12 +739,12 @@ class Block(PandasObject):
         mask = isna(values)
 
         if not self.is_object and not quoting:
-            itemsize = writers.max_len_string(na_rep)
-            values = values.astype("<U{size}".format(size=itemsize))
+            values[mask] = na_rep
+            values = values.astype(str)
         else:
             values = np.array(values, dtype='object')
+            values[mask] = na_rep
 
-        values[mask] = na_rep
         return values
 
     # block actions ####
```

Resulted in 

```
============================================= 93 failed, 45227 passed, 1185 skipped, 1000 xfailed, 22 xpassed, 120 warnings, 23 error in 456.06 seconds =============================================

```

I've got a few tests (19) that fail in my environment even when Travis CI build passes, so am pushing the second version to git. Is it similar to what you had in mind @jreback ?

Reply:
Looks like Travis had some tests failing after my last commit...